### DISCUSSION TITLE
Massive memory leak

### DISCUSSION THREAD

**Discussion by Northernside** [PrismarineJS/mineflayer#2779] [General]
Greetings,
me and one of my friends are currently working on a project where we collect UUIDs by 1st scanning servers from all public IP ranges and 2nd letting bots (through Mineflayer) join all servers we've collected.
This is my current core: https://paste.northernsi.de/famesedasa.js
However, the code takes more and more RAM as the minutes pass, I already experienced it taking 2 GiB of RAM after roughly 30 minutes. 
![image](https://user-images.githubusercontent.com/60970791/192115948-b8783a1c-0c07-440b-b7ff-3f0b802c800d.png)
As you can see in the screenshot, the RAM usage is going up in a rather constant way (after restarting the code, as you can see at the usage drop).

I'm not an experienced JS developer yet, so I'm happy if someone can explain me the (root) cause of this memory leak and maybe how to (basically) fix it ;D

**Comment by Northernside** [comment]
I also don't know if I should post it as an issue, because it might be my fault. Should I still open a issue from this discussion?

**Comment by CubeBeveled** [comment]
I'm having a similar issue with my chatbridge bot.
The longer it stays online the more the gc has to do stuff (higher cpu usage) and the more memory is used.

The bot spawns many times while its online and when it reconnects it cleans everything up (events and intervals). I even implemented a chunk and entity purging method

**Comment by cxvqo** [comment]
Maybe you should try using node-Minecraft-protocol directly, at least for the server scanning purpose, it’d be okay, you don’t really need physics on a scanner bot. Though for such a supposedly massive project, you should consider using something like MCProtocolLib on Java/Kotlin.
